libsolidity/Parser.cpp

```diff
@@ -598,48 +620,85 @@ ASTPointer<ForStatement> Parser::parseForStatement()
 												body);
 }
 
-ASTPointer<Statement> Parser::parseVarDeclOrExprStmt()
+ASTPointer<Statement> Parser::parseSimpleStatement()
 {
-	if (peekVariableDeclarationStatement())
+	// These two cases are very hard to distinguish:
+	// x[7 * 20 + 3] a;  -  x[7 * 20 + 3] = 9;
+	// In the first case, x is a type name, in the second it is the name of a variable.
+	int isVariableDeclaration = peekVariableDeclarationStatement();
+	if (isVariableDeclaration == 1)
 		return parseVariableDeclarationStatement();
-	else
+	else if (isVariableDeclaration == -1)
 		return parseExpressionStatement();
+
+	// At this point, we have '(Identifier|ElementaryTypeName) "["'.
+	// We parse '(Identifier|ElementaryTypeName) ( "[" Expression "]" )+' and then decide whether to hand this over
+	// to ExpressionStatement or create a VariableDeclarationStatement out of it.
+	ASTPointer<PrimaryExpression> primary;
+	if (m_scanner->getCurrentToken() == Token::Identifier)
+		primary = parseIdentifier();
+	else
+	{
+		primary = ASTNodeFactory(*this).createNode<ElementaryTypeNameExpression>(m_scanner->getCurrentToken());
+		m_scanner->next();
+	}
+	vector<pair<ASTPointer<Expression>, Location>> indices;
+	solAssert(m_scanner->getCurrentToken() == Token::LBrack, "");
+	Location indexLocation = primary->getLocation();
+	do
+	{
+		expectToken(Token::LBrack);
+		ASTPointer<Expression> index;
+		if (m_scanner->getCurrentToken() != Token::RBrack)
+			index = parseExpression();
+		indexLocation.end = getEndPosition();
+		indices.push_back(make_pair(index, indexLocation));
+		expectToken(Token::RBrack);
+	}
+	while (m_scanner->getCurrentToken() == Token::LBrack);
+
+	if (m_scanner->getCurrentToken() == Token::Identifier)
+		return parseVariableDeclarationStatement(typeNameFromArrayIndexStructure(primary, indices));
+	else
+		return parseExpressionStatement(expressionFromArrayIndexStructure(primary, indices));
 }
 
-ASTPointer<VariableDeclarationStatement> Parser::parseVariableDeclarationStatement()
+ASTPointer<VariableDeclarationStatement> Parser::parseVariableDeclarationStatement(
+	ASTPointer<TypeName> const& _lookAheadArrayType)
 {
-	ASTNodeFactory nodeFactory(*this);
 	VarDeclParserOptions options;
 	options.allowVar = true;
 	options.allowInitialValue = true;
-	ASTPointer<VariableDeclaration> variable = parseVariableDeclaration(options);
+	ASTPointer<VariableDeclaration> variable = parseVariableDeclaration(options, _lookAheadArrayType);
+	ASTNodeFactory nodeFactory(*this, variable);
 	return nodeFactory.createNode<VariableDeclarationStatement>(variable);
 }
 
-ASTPointer<ExpressionStatement> Parser::parseExpressionStatement()
+ASTPointer<ExpressionStatement> Parser::parseExpressionStatement(
+	ASTPointer<Expression> const& _lookAheadArrayExpression)
```

Comment:
Here too .. and well .. all of the other places where a small function signature is split into 2 lines.